auto-instrumentation/httpurlconnection/agent/src/main/java/io/opentelemetry/instrumentation/agent/httpurlconnection/HttpUrlConnectionPlugin.java

```diff
@@ -0,0 +1,195 @@
+                                                    "getHeaderFieldDate", String.class, Long.TYPE)))
+                            .replaceWith(
+                                    HttpUrlReplacements.class.getDeclaredMethod(
+                                            "replacementForHttpHeaderFieldDate",
+                                            HttpURLConnection.class,
+                                            String.class,
+                                            Long.TYPE))
+                            .method(
+                                    is(
+                                            HttpURLConnection.class.getDeclaredMethod(
+                                                    "getResponseCode")))
+                            .replaceWith(
+                                    HttpUrlReplacements.class.getDeclaredMethod(
+                                            "replacementForResponseCode", URLConnection.class))
+                            .method(
+                                    is(
+                                            HttpURLConnection.class.getDeclaredMethod(
+                                                    "getResponseMessage")))
+                            .replaceWith(
+                                    HttpUrlReplacements.class.getDeclaredMethod(
+                                            "replacementForResponseMessage", URLConnection.class))
+                            .method(is(URLConnection.class.getDeclaredMethod("getOutputStream")))
+                            .replaceWith(
+                                    HttpUrlReplacements.class.getDeclaredMethod(
+                                            "replacementForOutputStream", URLConnection.class))
+                            .method(is(URLConnection.class.getDeclaredMethod("getInputStream")))
+                            .replaceWith(
+                                    HttpUrlReplacements.class.getDeclaredMethod(
+                                            "replacementForInputStream", URLConnection.class))
+                            .method(is(HttpURLConnection.class.getDeclaredMethod("getErrorStream")))
+                            .replaceWith(
+                                    HttpUrlReplacements.class.getDeclaredMethod(
+                                            "replacementForErrorStream", HttpURLConnection.class))
+                            .on(isMethod()));
+        } catch (NoSuchMethodException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        // No operation.
+    }
+
+    @Override
+    public boolean matches(TypeDescription target) {
+        // TODO: Add configuration for package name
+        return target.isSamePackage(
+                new TypeDescription.ForPackageDescription(
+                        new PackageDescription.Simple("com.example.httpurlconnection")));
```

Comment:
@breedx-splk @LikeTheSalad Is there a better way to specify the packages to instrument here? 

1. I thought of using Context.getPackageName() but that could be different from the actual package names in the application code. 
2.  Just returning true for all types, to let byte buddy look at all .classes is very inefficient and it also looks into attached android libraries and was throwing errors for some kotlin libraries. 
3.  Best way I could think of was to allow customers to specify a list of package names from their application or may be the one topmost package name. This helps avoid any unnecessary instrumentations as well from any added third party libraries ( as we do not want to instrument other http client libraries that might internally use HttpURLConnection). May be a property like "httpURLConnectionPackage" specified somewhere in their application code property files rather than having them to make an API call to our APIs to set this?

Reply:
For targeting only classes that belong to the host project itself (and ignoring classes from libraries) the best option here would be to use the `AndroidDescriptor` tool. This is an example of how to do so: https://github.com/elastic/apm-agent-android/blob/main/android-instrumentation/src/main/java/co/elastic/apm/android/instrumentation/ui/activities/ActivityLifecyclePlugin.java#L50 

That way it's not needed to check for package names since the ByteBuddy gradle plugin already can tell you which classes are local, this is one of the benefits of instrumenting bytecode at compile-time.

Reply:
Thanks for suggesting this! It worked :)